### PR TITLE
HBASE-23116: Fix a couple of load balancer logging nits.

### PR DESCRIPTION
**Pull request by bharathv**
Fixes a typo and adds table name to the logging context when
StochasticLoadBalancer is configured "per table". Added some
test coverage with per-table balancer enabled and manually
verified the logs to make sure the table name is formatted
correctly.

### CODE REVIEW

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java [hunk 1]
```diff
@@ -1746,7 +1746,7 @@ public boolean balance(boolean force) throws IOException {
           toPrint = regionsInTransition.subList(0, max);
           truncated = true;
         }
-        LOG.info(prefix + "unning balancer because " + regionsInTransition.size() +
+        LOG.info(prefix + "Running balancer because " + regionsInTransition.size() +
```
Comment (virajjasani) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java]: Ideally, this change is not required since prefix takes care of it. 
However, to improve text searching, we can make this change:
```
String prefix = force && !metaInTransition ? "" : "Not ";
```

File: hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestStochasticLoadBalancer.java [hunk 2]
```diff
@@ -35,13 +35,7 @@
...
 
 import org.apache.hadoop.conf.Configuration;
-import org.apache.hadoop.hbase.ClusterMetrics;
-import org.apache.hadoop.hbase.HBaseClassTestRule;
-import org.apache.hadoop.hbase.HBaseConfiguration;
-import org.apache.hadoop.hbase.RegionMetrics;
-import org.apache.hadoop.hbase.ServerMetrics;
-import org.apache.hadoop.hbase.ServerName;
-import org.apache.hadoop.hbase.Size;
+import org.apache.hadoop.hbase.*;
```
Comment (virajjasani) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestStochasticLoadBalancer.java]: Please revert this change

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/StochasticLoadBalancer.java [hunk 3]
```diff
@@ -348,9 +348,10 @@ protected boolean needsBalance(Cluster cluster) {
     if (total <= 0 || sumMultiplier <= 0
         || (sumMultiplier > 0 && (total / sumMultiplier) < minCostNeedBalance)) {
       if (LOG.isTraceEnabled()) {
-        LOG.trace("Skipping load balancing because balanced cluster; " + "total cost is " + total
-          + ", sum multiplier is " + sumMultiplier + " min cost which need balance is "
-          + minCostNeedBalance);
+        final String loadBalanceTarget = isByTable ? String.format("table (%s)", tableName) : "cluster";
```
Comment (virajjasani) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/StochasticLoadBalancer.java]: nit: although loadBalanceTarget is used just for logging, it is better to keep only log.trace() within LOG.isTraceEnabled()

File: hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestStochasticLoadBalancer.java [hunk 4]
```diff
@@ -242,15 +236,24 @@ public void testKeepRegionLoad() throws Exception {
...
     float minCost = conf.getFloat("hbase.master.balancer.stochastic.minCostNeedBalance", 0.05f);
     conf.setFloat("hbase.master.balancer.stochastic.minCostNeedBalance", 1.0f);
-    loadBalancer.setConf(conf);
-    for (int[] mockCluster : clusterStateMocks) {
-      Map<ServerName, List<RegionInfo>> servers = mockClusterServers(mockCluster);
-      List<RegionPlan> plans = loadBalancer.balanceCluster(servers);
-      assertNull(plans);
+    try {
+      // Test with/without per table balancer.
+      boolean[] perTableBalancerConfigs = {true, false};
```
Comment (virajjasani) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestStochasticLoadBalancer.java]: make it private static final at class level?

File: hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestStochasticLoadBalancer.java [hunk 5]
```diff
@@ -242,15 +236,24 @@ public void testKeepRegionLoad() throws Exception {
...
+        loadBalancer.setConf(conf);
+        for (int[] mockCluster : clusterStateMocks) {
+          Map<ServerName, List<RegionInfo>> servers = mockClusterServers(mockCluster);
+          List<RegionPlan> plans = loadBalancer.balanceCluster(servers);
+          assertNull(plans);
+        }
+      }
+    } finally {
+      // reset config
+      conf.setBoolean(HConstants.HBASE_MASTER_LOADBALANCE_BYTABLE, false);
```
Comment (virajjasani) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestStochasticLoadBalancer.java]: if intention is to revert applied configs, we can unset this config

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   0m 37s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   7m 12s |  master passed  |
| :green_heart: |  compile  |   1m  0s |  master passed  |
| :green_heart: |  checkstyle  |   1m 29s |  master passed  |
| :green_heart: |  shadedjars  |   4m 49s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 41s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 39s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 35s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   6m 31s |  the patch passed  |
| :green_heart: |  compile  |   1m 10s |  the patch passed  |
| :green_heart: |  javac  |   1m 10s |  the patch passed  |
| :broken_heart: |  checkstyle  |   1m 34s |  hbase-server: The patch generated 2 new + 170 unchanged - 0 fixed = 172 total (was 170)  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   5m 39s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  19m 25s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 39s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 58s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 168m 40s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 34s |  The patch does not generate ASF License warnings.  |
|  |   | 237m 19s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.security.access.TestSnapshotScannerHDFSAclController |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-687/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/687 |
| JIRA Issue | HBASE-23116 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 977c3daa96bb 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-687/out/precommit/personality/provided.sh |
| git revision | master / 5aa8d3a20b |
| Default Java | 1.8.0_181 |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-687/1/artifact/out/diff-checkstyle-hbase-server.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-687/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-687/1/testReport/ |
| Max. process+thread count | 4480 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-687/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java (hunk 1 above)
Comment (wchevreuil) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java]: +1 to @virajjasani  comment.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/StochasticLoadBalancer.java (hunk 3 above)
Comment (wchevreuil) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/StochasticLoadBalancer.java]: @virajjasani better to keep like this, as we don't need the extra execution in case TRACE isn't on, and keeping it as a separate variable gives better readability.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestStochasticLoadBalancer.java (hunk 4 above)
Comment (wchevreuil) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestStochasticLoadBalancer.java]: Should be fine as local, since it's only used at this method scope.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/StochasticLoadBalancer.java (hunk 3 above)
Comment (virajjasani) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/StochasticLoadBalancer.java]: Definitely keeping it as a variable is fine. I just thought it would be better to keep it out of  `LOG.isTraceEnabled()` since usually we want only `LOG.trace()` to be covered by it. But anyways, keeping it unchanged is fine, it's just used by trace log.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java (hunk 1 above)
Comment (bharathv) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java]: Ouch, haven't realized that prefix. Will revert this piece. I believe it exists this way for appropriate capitalization, so don't want to break it.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestStochasticLoadBalancer.java (hunk 4 above)
Comment (bharathv) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestStochasticLoadBalancer.java]: Yes, this is very specific to the test, don't think should be at the class scope.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestStochasticLoadBalancer.java (hunk 5 above)
Comment (bharathv) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestStochasticLoadBalancer.java]: Done.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestStochasticLoadBalancer.java (hunk 2 above)
Comment (bharathv) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestStochasticLoadBalancer.java]: Dang, intellij! Fixed.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/StochasticLoadBalancer.java [hunk 6]
```diff
@@ -348,9 +348,11 @@ protected boolean needsBalance(Cluster cluster) {
     if (total <= 0 || sumMultiplier <= 0
         || (sumMultiplier > 0 && (total / sumMultiplier) < minCostNeedBalance)) {
       if (LOG.isTraceEnabled()) {
-        LOG.trace("Skipping load balancing because balanced cluster; " + "total cost is " + total
-          + ", sum multiplier is " + sumMultiplier + " min cost which need balance is "
-          + minCostNeedBalance);
+        final String loadBalanceTarget =
+            isByTable ? String.format("table (%s)", tableName) : "cluster";
+        LOG.trace("Skipping load balancing because the {} is balanced. Total cost: {}, "
```
Comment (virajjasani) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/StochasticLoadBalancer.java]: nit: from the log msg, "the" can be removed. `the cluster is balanced` is fine but `the t1 is balanced`?

**Comment by wchevreuil** [comment]
LGTM, +1. Also ran this previously failed TestSnapshotScannerHDFSAclController and it's passing locally, so it should be unrelated. Will merge this PR.

**Comment by apurtell** [comment]
There are fix versions for 1.x on the JIRA but no commits to branch-1s. Is there a backport in progress?

**Comment by apurtell** [comment]
Answering my own question - I am doing the backport now. Trying to get 1.5.0 RC out.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   1m 10s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 21s |  master passed  |
| :green_heart: |  compile  |   0m 56s |  master passed  |
| :green_heart: |  checkstyle  |   1m 21s |  master passed  |
| :green_heart: |  shadedjars  |   4m 36s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 37s |  master passed  |
| :blue_heart: |  spotbugs  |   4m  3s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m  1s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   4m 58s |  the patch passed  |
| :green_heart: |  compile  |   0m 53s |  the patch passed  |
| :green_heart: |  javac  |   0m 53s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 19s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m 34s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  15m 33s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 37s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 25s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 307m 57s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 38s |  The patch does not generate ASF License warnings.  |
|  |   | 365m 17s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.client.TestSnapshotCloneIndependence |
|   | hadoop.hbase.regionserver.TestRegionReplicasWithModifyTable |
|   | hadoop.hbase.master.TestSplitWALManager |
|   | hadoop.hbase.client.TestSnapshotTemporaryDirectoryWithRegionReplicas |
|   | hadoop.hbase.client.TestAsyncTableAdminApi |
|   | hadoop.hbase.client.TestFromClientSide |
|   | hadoop.hbase.client.TestFromClientSideWithCoprocessor |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.2 Server=19.03.2 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-687/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/687 |
| JIRA Issue | HBASE-23116 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 3210a8a8ed2d 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-687/out/precommit/personality/provided.sh |
| git revision | master / 800c35a30e |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-687/2/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-687/2/testReport/ |
| Max. process+thread count | 4782 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-687/2/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.
